### ISSUE TITLE
InvalidOperationException when calling GetData on IDataObject

### ISSUE THREAD

**Issue by livarcocc** [dotnet/corefx#33616]
@nick-beer commented on [Fri Nov 16 2018](https://github.com/dotnet/core-sdk/issues/156)

In porting our test suite to .NET Core 3.0, I'm seeing a new failure in one of our tests.  When trying to call the GetData API on an instance of IDataObject returned from Clipboard.GetDataObject, I get an InvalidOperationException with the following message:

*This type has a ComVisible(false) parent in its hierarchy, therefore QueryInterface calls for IDispatch or class interfaces are disallowed*

This may be related to issue #149

To reproduce the issue:
- clone the repo https://github.com/nick-beer/IDataObjectBug
- Run the test *DataObject_Fails* from within the 'Tests' folder

**Comment by rladuca** [comment]
@livarcocc There seems to be a difference in the interface definitions between core and desktop framework.

In desktop, IDataObject is [decorated with COM attributes](https://referencesource.microsoft.com/#System/compmod/System/Runtime/InteropServices/ComTypes/IDataObject.cs,541e28add80cf992).

In core, `[it is not](https://source.dot.net/#System.Runtime.InteropServices/System.Runtime.InteropServices.cs,541e28add80cf992)`.

**Comment by ericstj** [comment]
@rladuca are you sure?
https://github.com/dotnet/corefx/blob/c3154a9642a92aa15704abce7bf695b3e9db55f5/src/System.Runtime.InteropServices/src/System/Runtime/InteropServices/ComTypes/IDataObject.cs#L18-L21

Be careful with source.dot.net, that doesn't handle reference assemblies very well.

Moreover, the IDataObject used in this repro is actually defined by PresentationCore.
src/WPF/src/PresentationCore/CSharp/System/Windows/IDataObject.cs

Here's a simpler repro:
```C#
using System;
using System.Windows;

namespace testClipboard
{
    class Program
    {
        [STAThread]
        static void Main(string[] args)
        {
            var data = new TestClipboardData("ABC");
            var dataObject = new DataObject();
            dataObject.SetData(data);
            Clipboard.SetDataObject(dataObject);

            IDataObject clipboardData = Clipboard.GetDataObject();           

            var testClipboardData = clipboardData.GetData(typeof(TestClipboardData)) as TestClipboardData;
            
        }

        [Serializable]
        internal class TestClipboardData
        {
            public string Text { get; private set; }

            public TestClipboardData(string text)
            {
                Text = text;
            }
        }
    }
}

```

**Comment by ericstj** [comment]
@jeffschwMSFT this is happening because System.Type is not COMVisible in .NETCore.

Winforms has nearly the same code but doesn't hit this because they [wrap](https://referencesource.microsoft.com/#System.Windows.Forms/winforms/Managed/System/WinForms/Clipboard.cs,257) the COM object with a managed object that will always translate the GetObject call to use a [string](https://referencesource.microsoft.com/#System.Windows.Forms/winforms/Managed/System/WinForms/DataObject.cs,210) instead of type.

WPF on the other hand, [does not wrap](https://referencesource.microsoft.com/#PresentationCore/Core/CSharp/System/Windows/Clipboard.cs,716) the COM object so it returns the raw interface.  When the test code calls GetObject(type) that's what triggers the runtime to complain about the COM visibility of System.Type.

To workaround the caller can avoid calling GetObject(type) and instead just call GetObject(type.FullName).

To fix this (if the runtime doesn't add back COMVisibility to System.Type) WPF can wrap with a managed object like WinForms does, but I think the better fix would be to add back COMVisible to System.Type.  @jkotas / @jeffschwMSFT any reason not to do that?

**Comment by jeffschwMSFT** [comment]
cc @AaronRobinsonMSFT

**Comment by jkotas** [comment]
> I think the better fix would be to add back COMVisible to System.Type

We are not going to make any of the regular managed types ComVisible. It is even worse compatibility nightmare than binary serialization. The workaround you have suggested is the right way to fix this.

**Comment by ericstj** [comment]
@AaronRobinsonMSFT are you talking about this: https://github.com/dotnet/corefx/issues/31079#issuecomment-406037340

I agree that we can make a tactical fix in WPF here, but that's just because we have another interface member we can call to get the same job done without relying on Type being COMVisible.  I can imagine that the full extent of this break could be rather difficult to track down and not always fixable.  At the very least we should update the exception message to let folks know what type the runtime is complaining about and document the extent of the types that we've removed COMVisible from.  It may also be worthwhile to scan all the interfaces we use with COM to see where else we're relying on deprecated COM visibility.

**Comment by ericstj** [comment]
https://github.com/dotnet/corefx/issues/33638 is another issue caused by Type not being COMVisible.

**Comment by danmosemsft** [comment]
> member we can call to get the same job done without relying on Type being COMVisible. I can imagine that the full extent of this break could be rather difficult to track down and not always fixable. At the very least we should update the exception message to let folks know what type the runtime is complaining abou

@AaronRobinsonMSFT thoughts on the above? It seems this is what remains of this issue.

**Comment by danmosemsft** [comment]
@AaronRobinsonMSFT I'd like to get consensus on what if anything to do here so we can get it to happen for 3.0 if we need to..?

**Comment by danmosemsft** [comment]
@AaronRobinsonMSFT given your comment above, shall we move to CoreCLR? I think it makes a lot of sense to do this in 3.0 if it will help customers unblock themselves without having to reach out to us.

**Comment by AaronRobinsonMSFT** [comment]
@danmosemsft Yep. Moving this to CoreCLR makes sense. Is there a way to move the issue across repos so we don't lose context?
